File: packages/sdk-python/src/stagehand/stagehand.py
from __future__ import annotations

import asyncio
import builtins
import inspect
import json
import sys
from collections.abc import Callable, Mapping, Sequence
from importlib.metadata import version
from pathlib import Path
from types import TracebackType
from typing import Literal, Self, TypeVar, overload

from pydantic import BaseModel

from ._generated.models import (
    Action,
    ActOptions,
    ActResult,
    BrowserbaseBrowserSettings,
    BrowserbaseProxyConfig,
    BrowserbaseRegion,
    ClientModelReference,
    EmptyParams,
    ExternalProxyConfig,
    ExtractOptions,
    ImplementationInfo,
    LLMGenerateParams,
    LLMGenerateResult,
    ModelConfig,
    ObserveOptions,
    ObserveResult,
    ProxyConfig,
    StagehandActParams,
    StagehandCloseResult,
    StagehandExtractParams,
    StagehandInitParams,
    StagehandInitResult,
    StagehandLog,
    StagehandMetrics,
    StagehandObserveParams,
    TelemetryConfig,
    Variables,
)
from ._generated.models import (
    Locator as ProtocolLocator,
)
from ._generated.protocol_version import STAGEHAND_PROTOCOL_VERSION
from .browser_context import BrowserContext
from .browser_source import ResolvedBrowserSource, resolve_browser_source
from .cdp_client import CDPConnectionClosedError
from .client_models import (
    BrowserbaseBrowserSource,
    Cache,
    CdpBrowserSource,
    ClientLLM,
    ExtractResult,
    LLMGenerateCallback,
    LocalBrowserSource,
    LocalProxyConfig,
    LocalViewport,
    StagehandClientInitParams,
    StagehandClientLoggingConfig,
    _cache_config,
    _ExtractWireResult,
    _model_config,
)
from .page import Page
from .rpc_client import RPCClient, connect_rpc_client

ResultModel = TypeVar("ResultModel", bound=BaseModel)


class Stagehand:
    @overload
    def __init__(
        self,
        *,
        browser: Literal["local"],
        api_key: str | None = None,
        args: Sequence[str] | None = None,
        executable_path: str | Path | None = None,
        port: int | None = None,
        user_data_dir: str | Path | None = None,
        preserve_user_data_dir: bool | None = None,
        headless: bool | None = None,
        devtools: bool | None = None,
        chromium_sandbox: bool | None = None,
        ignore_default_args: bool | Sequence[str] | None = None,
        proxy_server: str | None = None,
        proxy_bypass: str | None = None,
        proxy_username: str | None = None,
        proxy_password: str | None = None,
        locale: str | None = None,
        viewport_width: int | None = None,
        viewport_height: int | None = None,
        device_scale_factor: float | None = None,
        has_touch: bool | None = None,
        ignore_https_errors: bool | None = None,
        connect_timeout_ms: int | None = None,
        downloads_path: str | Path | None = None,
        accept_downloads: bool | None = None,
        keep_alive: bool | None = None,
        model: str | LLMGenerateCallback | None = None,
        model_api_key: str | None = None,
        model_base_url: str | None = None,
        model_headers: Mapping[str, str] | None = None,
        telemetry: TelemetryConfig | None = None,
        system_prompt: str | None = None,
        self_heal: bool | None = None,
        dom_settle_timeout_ms: int | None = None,
        cache: Cache | None = None,
        logging: StagehandClientLoggingConfig | None = None,
    ) -> None: ...

    @overload
    def __init__(
        self,
        *,
        browser: Literal["cdp"],
        cdp_url: str,
        headers: Mapping[str, str] | None = None,
        api_key: str | None = None,
        model: str | LLMGenerateCallback | None = None,
        model_api_key: str | None = None,
        model_base_url: str | None = None,
        model_headers: Mapping[str, str] | None = None,
        telemetry: TelemetryConfig | None = None,
        system_prompt: str | None = None,
        self_heal: bool | None = None,
        dom_settle_timeout_ms: int | None = None,
        cache: Cache | None = None,
        logging: StagehandClientLoggingConfig | None = None,
    ) -> None: ...

    @overload
    def __init__(
        self,
        *,
        api_key: str,
        browser: Literal["browserbase"] = "browserbase",
        browser_settings: BrowserbaseBrowserSettings | None = None,
        extension_id: str | None = None,
        keep_alive: bool | None = None,
        proxies: bool | Sequence[BrowserbaseProxyConfig | ExternalProxyConfig] | None = None,
        region: BrowserbaseRegion
        | Literal["us-west-2", "us-east-1", "eu-central-1", "ap-southeast-1"]
        | None = None,
        timeout: float | None = None,
        user_metadata: Mapping[str, object] | None = None,
        model: str | LLMGenerateCallback | None = None,
        model_api_key: str | None = None,
        model_base_url: str | None = None,
        model_headers: Mapping[str, str] | None = None,
        telemetry: TelemetryConfig | None = None,
        system_prompt: str | None = None,
        self_heal: bool | None = None,
        dom_settle_timeout_ms: int | None = None,
        cache: Cache | None = None,
        logging: StagehandClientLoggingConfig | None = None,
    ) -> None: ...

    def __init__(
        self,
        *,
        browser: Literal["browserbase", "local", "cdp"] = "browserbase",
        api_key: str | None = None,
        browser_settings: BrowserbaseBrowserSettings | None = None,
        extension_id: str | None = None,
        proxies: bool | Sequence[BrowserbaseProxyConfig | ExternalProxyConfig] | None = None,
        region: BrowserbaseRegion
        | Literal["us-west-2", "us-east-1", "eu-central-1", "ap-southeast-1"]
        | None = None,
        timeout: float | None = None,
        user_metadata: Mapping[str, object] | None = None,
        args: Sequence[str] | None = None,
        executable_path: str | Path | None = None,
        port: int | None = None,
        user_data_dir: str | Path | None = None,
        preserve_user_data_dir: bool | None = None,
        headless: bool | None = None,
        devtools: bool | None = None,
        chromium_sandbox: bool | None = None,
        ignore_default_args: bool | Sequence[str] | None = None,
        proxy_server: str | None = None,
        proxy_bypass: str | None = None,
        proxy_username: str | None = None,
        proxy_password: str | None = None,
        locale: str | None = None,
        viewport_width: int | None = None,
        viewport_height: int | None = None,
        device_scale_factor: float | None = None,
        has_touch: bool | None = None,
        ignore_https_errors: bool | None = None,
        connect_timeout_ms: int | None = None,
        downloads_path: str | Path | None = None,
        accept_downloads: bool | None = None,
        keep_alive: bool | None = None,
        cdp_url: str | None = None,
        headers: Mapping[str, str] | None = None,
        model: str | LLMGenerateCallback | None = None,
        model_api_key: str | None = None,
        model_base_url: str | None = None,
        model_headers: Mapping[str, str] | None = None,
        telemetry: TelemetryConfig | None = None,
        system_prompt: str | None = None,
        self_heal: bool | None = None,
        dom_settle_timeout_ms: int | None = None,
        cache: Cache | None = None,
        logging: StagehandClientLoggingConfig | None = None,
    ) -> None:
        if browser == "browserbase":
            browser_source = BrowserbaseBrowserSource.model_validate({
                "type": "browserbase",
                **{
                    name: value
                    for name, value in (
                        ("browser_settings", browser_settings),
                        ("extension_id", extension_id),
                        ("keep_alive", keep_alive),
                        (
                            "proxies",
                            (
                                proxies
                                if isinstance(proxies, bool) or proxies is None
                                else [ProxyConfig(root=proxy) for proxy in proxies]
                            ),
                        ),
                        ("region", region),
                        ("timeout", timeout),
                        (
                            "user_metadata",
                            dict(user_metadata) if user_metadata is not None else None,
                        ),
                    )
                    if value is not None
                },
            })
        elif browser == "local":
            if (viewport_width is None) != (viewport_height is None):
                raise TypeError("viewport_width and viewport_height must be provided together")
            if proxy_server is None and any(
                value is not None for value in (proxy_bypass, proxy_username, proxy_password)
            ):
                raise TypeError("proxy_server is required when configuring a local proxy")
            browser_source = LocalBrowserSource.model_validate({
                "type": "local",
                **{
                    name: value
                    for name, value in (
                        ("args", list(args) if args is not None else None),
                        (
                            "executable_path",
                            str(executable_path) if executable_path is not None else None,
                        ),
                        ("port", port),
                        (
                            "user_data_dir",
                            str(user_data_dir) if user_data_dir is not None else None,
                        ),
                        ("preserve_user_data_dir", preserve_user_data_dir),
                        ("headless", headless),
                        ("devtools", devtools),
                        ("chromium_sandbox", chromium_sandbox),
                        (
                            "ignore_default_args",
                            (
                                list(ignore_default_args)
                                if not isinstance(ignore_default_args, bool)
                                and ignore_default_args is not None
                                else ignore_default_args
                            ),
                        ),
                        (
                            "proxy",
                            (
                                LocalProxyConfig(
                                    server=proxy_server,
                                    bypass=proxy_bypass,
                                    username=proxy_username,
                                    password=proxy_password,
                                )
                                if proxy_server is not None
                                else None
                            ),
                        ),
                        ("locale", locale),
                        (
                            "viewport",
                            (
                                LocalViewport(width=viewport_width, height=viewport_height)
                                if viewport_width is not None and viewport_height is not None
                                else None
                            ),
                        ),
                        ("device_scale_factor", device_scale_factor),
                        ("has_touch", has_touch),
                        ("ignore_https_errors", ignore_https_errors),
                        ("connect_timeout_ms", connect_timeout_ms),
                        (
                            "downloads_path",
                            str(downloads_path) if downloads_path is not None else None,
                        ),
                        ("accept_downloads", accept_downloads),
                        ("keep_alive", keep_alive),
                    )
                    if value is not None
                },
            })
        elif browser == "cdp":
            if cdp_url is None:
                raise TypeError("cdp_url is required when browser='cdp'")
            browser_source = CdpBrowserSource(
                type="cdp",
                cdp_url=cdp_url,
                **({"headers": dict(headers)} if headers is not None else {}),
            )
        else:
            raise ValueError(f"Unsupported browser source: {browser}")

        model_connection_options = (model_api_key, model_base_url, model_headers)
        if model is None and any(value is not None for value in model_connection_options):
            raise TypeError("model connection options require a model name")
        if callable(model) and any(value is not None for value in model_connection_options):
            raise TypeError("model connection options cannot be used with an LLM callback")

        resolved_model: ModelConfig | ClientLLM | None
        if isinstance(model, str):
            resolved_model = _model_config(
                model,
                api_key=model_api_key,
                base_url=model_base_url,
                headers=dict(model_headers) if model_headers is not None else None,
            )
        elif model is not None:
            resolved_model = ClientLLM(generate=model)
        else:
            resolved_model = None

        values: dict[str, object] = {
            name: value
            for name, value in (
                ("api_key", api_key),
                ("system_prompt", system_prompt),
                ("self_heal", self_heal),
                ("dom_settle_timeout_ms", dom_settle_timeout_ms),
                ("cache", _cache_config(cache) if cache is not None else None),
                ("logging", logging),
            )
            if value is not None
        }
        values["browser"] = browser_source
        if resolved_model is not None:
            values["model"] = resolved_model
        if telemetry is not None:
            values["telemetry"] = telemetry
        self.init_params = StagehandClientInitParams.model_validate(values)
        self._browser_context: BrowserContext | None = None
        self._rpc_client: RPCClient | None = None
        self._remove_notification_listener: Callable[[], None] | None = None
        self._remove_client_llm_handler: Callable[[], None] | None = None
        self._browser: ResolvedBrowserSource | None = None
        self._initialized = False
        self._lifecycle_lock = asyncio.Lock()

    @property
    def context(self) -> BrowserContext:
        if self._browser_context is None:
            raise RuntimeError(
                "Stagehand is not initialized. Call stagehand.init() before using context."
            )
        return self._browser_context

    @property
    def browser(self) -> ResolvedBrowserSource:
        if self._browser is None:
            raise RuntimeError(
                "Stagehand is not initialized. Call stagehand.init() before using browser."
            )
        return self._browser

    @property
    def initialized(self) -> bool:
        return self._initialized

    async def metrics(self) -> StagehandMetrics:
        return await self._connected_rpc_client.send(
            "stagehand.metrics",
            EmptyParams(),
            StagehandMetrics,
        )

    async def init(self) -> None:
        async with self._lifecycle_lock:
            if self._initialized:
                return

            browser = await resolve_browser_source(self.init_params)
            self._browser = browser
            extension_dir = Path(__file__).with_name("_extension")
            if not (extension_dir / "manifest.json").is_file():
                extension_dir = Path(__file__).resolve().parents[3] / "server" / "dist"

            try:
                rpc_client = await connect_rpc_client(
                    cdp_url=browser.cdp_url,
                    extension_dir=str(extension_dir),
                    service_worker_url_includes="service-worker.js",
                    cdp_connect_timeout_ms=browser.connect_timeout_ms or 10_000,
                )
                self._rpc_client = rpc_client
                self._remove_notification_listener = rpc_client.on_notification(
                    "stagehand.log",
                    StagehandLog,
                    self._handle_stagehand_notification,
                )
                client_llm = self.init_params.model
                if isinstance(client_llm, ClientLLM):

                    async def generate(params: LLMGenerateParams) -> LLMGenerateResult:
                        return LLMGenerateResult(root=await client_llm.generate(params.root))

                    self._remove_client_llm_handler = rpc_client.on_request(
                        "llm.generate",
                        LLMGenerateParams,
                        LLMGenerateResult,
                        generate,
                    )

                browser_cdp_url = rpc_client.browser_web_socket_debugger_url
                if not browser.resident_browser_connection and browser_cdp_url is None:
                    raise RuntimeError("The browser CDP WebSocket URL is unavailable")
                await rpc_client.send(
                    "stagehand.init",
                    self._worker_init_params(
                        None if browser.resident_browser_connection else browser_cdp_url
                    ),
                    StagehandInitResult,
                )
                self._browser_context = BrowserContext(rpc_client)
            except BaseException:
                await asyncio.shield(self._release_resources())
                raise

            self._initialized = True

    async def act(
        self,
        instruction: str | Action,
        *,
        page: Page | None = None,
        model: ModelConfig | None = None,
        variables: Variables | None = None,
        timeout: float | None = None,
        locator: ProtocolLocator | None = None,
        cache: Cache | None = None,
    ) -> ActResult:
        options = ActOptions.model_validate({
            name: value
            for name, value in (
                ("model", model),
                ("variables", variables),
                ("timeout", timeout),
                ("locator", locator),
                ("cache", _cache_config(cache) if cache is not None else None),
            )
            if value is not None
        })
        target_page = page or await self.context.active_page()
        if target_page is None:
            raise RuntimeError("Stagehand has no active page")
        params = StagehandActParams.model_validate({
            "page_id": target_page.page_id,
            "instruction": instruction,
        })
        if options.model_fields_set:
            params.options = options
        result = await self._connected_rpc_client.send("stagehand.act", params, ActResult)
        return result

    async def observe(
        self,
        *,
        instruction: str | None = None,
        page: Page | None = None,
        model: ModelConfig | None = None,
        variables: Variables | None = None,
        timeout: float | None = None,
        selector: str | None = None,
        ignore_selectors: list[str] | None = None,
        locator: ProtocolLocator | None = None,
        cache: Cache | None = None,
    ) -> ObserveResult:
        options = ObserveOptions.model_validate({
            name: value
            for name, value in (
                ("model", model),
                ("variables", variables),
                ("timeout", timeout),
                ("selector", selector),
                ("ignore_selectors", ignore_selectors),
                ("locator", locator),
                ("cache", _cache_config(cache) if cache is not None else None),
            )
            if value is not None
        })
        target_page = page or await self.context.active_page()
        if target_page is None:
            raise RuntimeError("Stagehand has no active page")
        params = StagehandObserveParams(page_id=target_page.page_id, instruction=instruction)
        if options.model_fields_set:
            params.options = options
        result = await self._connected_rpc_client.send("stagehand.observe", params, ObserveResult)
        return result

    async def extract(
        self,
        *,
        instruction: str,
        schema: builtins.type[ResultModel],
        page: Page | None = None,
        model: ModelConfig | None = None,
        timeout: float | None = None,
        selector: str | None = None,
        ignore_selectors: list[str] | None = None,
        screenshot: bool | None = None,
        locator: ProtocolLocator | None = None,
        cache: Cache | None = None,
    ) -> ExtractResult[ResultModel]:
        options = ExtractOptions.model_validate({
            name: value
            for name, value in (
                ("model", model),
                ("timeout", timeout),
                ("selector", selector),
                ("ignore_selectors", ignore_selectors),
                ("screenshot", screenshot),
                ("locator", locator),
                ("cache", _cache_config(cache) if cache is not None else None),
            )
            if value is not None
        })
        target_page = page or await self.context.active_page()
        if target_page is None:
            raise RuntimeError("Stagehand has no active page")
        params = StagehandExtractParams(
            page_id=target_page.page_id,
            instruction=instruction,
            schema_=schema.model_json_schema(),
        )
        if options.model_fields_set:
            params.options = options
        result = await self._connected_rpc_client.send(
            "stagehand.extract", params, _ExtractWireResult
        )
        return ExtractResult(
            data=schema.model_validate(result.data),
            metadata=result.metadata,
        )

    async def close(self) -> None:
        async with self._lifecycle_lock:
            try:
                if self._browser_context is not None and self._rpc_client is not None:
                    try:
                        await self._rpc_client.send(
                            "stagehand.close",
                            EmptyParams(),
                            StagehandCloseResult,
                        )
                    except CDPConnectionClosedError:
                        pass
            finally:
                await asyncio.shield(self._release_resources())

    async def __aenter__(self) -> Self:
        await self.init()
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_value: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        await self.close()

    @property
    def _connected_rpc_client(self) -> RPCClient:
        if not self._initialized or self._rpc_client is None:
            raise RuntimeError(
                "Stagehand is not initialized. Call stagehand.init() before using it."
            )
        return self._rpc_client

    def _worker_init_params(self, browser_cdp_url: str | None) -> StagehandInitParams:
        values = self.init_params.model_dump(
            exclude={"browser", "logging", "model"},
            exclude_unset=True,
        )
        if isinstance(self.init_params.model, ClientLLM):
            values["model"] = ClientModelReference(source="client")
        elif self.init_params.model is not None:
            values["model"] = self.init_params.model
        values["protocol_version"] = STAGEHAND_PROTOCOL_VERSION
        values["client_info"] = ImplementationInfo(
            name="stagehand-sdk-python",
            version=version("stagehand"),
        )
        values["log_level"] = self.init_params.logging.level
        if browser_cdp_url is not None:
            values["browser_cdp_url"] = browser_cdp_url
        return StagehandInitParams.model_validate(values)

    async def _handle_stagehand_notification(self, notification: StagehandLog) -> None:
        logging = self.init_params.logging
        if not _is_log_level_enabled(notification.level.value, logging.level):
            return

        sys.stderr.write(_render_stagehand_log(notification, logging.format) + "\n")
        if logging.on_log is None:
            return

        try:
            result = logging.on_log(notification)
            if inspect.isawaitable(result):
                await result
        except Exception as error:
            sys.stderr.write(f"[stagehand] ERROR on_log callback failed: {error}\n")

    async def _release_resources(self) -> None:
        if self._remove_client_llm_handler is not None:
            self._remove_client_llm_handler()
            self._remove_client_llm_handler = None
        if self._remove_notification_listener is not None:
            self._remove_notification_listener()
            self._remove_notification_listener = None
        rpc_client = self._rpc_client
        self._rpc_client = None
        browser = self._browser
        self._browser = None
        self._browser_context = None
        self._initialized = False
        try:
            if rpc_client is not None:
                await rpc_client.close()
        finally:
            if browser is not None and not browser.keep_alive:
                await browser.close()


_LOG_LEVEL_PRIORITY = {
    "debug": 10,
    "info": 20,
    "warn": 30,
    "error": 40,
    "off": float("inf"),
}


def _is_log_level_enabled(level: str, threshold: str) -> bool:
    return _LOG_LEVEL_PRIORITY[level] >= _LOG_LEVEL_PRIORITY[threshold]


def _render_stagehand_log(notification: StagehandLog, format_: str) -> str:
    data = notification.data.model_dump(mode="json")
    record = {
        "level": notification.level.value,
        "message": notification.message,
        "data": data,
    }
    if format_ == "json":
        return json.dumps(record, separators=(",", ":"))

    suffix = "" if not data else f" {json.dumps(data, separators=(',', ':'))}"
    return f"[stagehand] {notification.level.value.upper()} {notification.message}{suffix}"

File: packages/sdk-ts/src/stagehand.ts
import { connectRPCClient, type RPCClient, type RPCClientOptions } from "./rpcClient.js";
import { STAGEHAND_PROTOCOL_VERSION, StagehandInitParamsSchema } from "../../protocol/schemas.js";
import { StagehandMethods } from "../../protocol/schema-registry.js";
import type {
  Action,
  ActResult,
  ObserveResult,
  StagehandMetrics,
  StagehandRpcNotification,
} from "../../protocol/types.js";
import { z } from "zod/v4";
import { BrowserContext } from "./browserContext.js";
import { resolveBrowserSource, type ResolvedBrowserSource } from "./browserSource.js";
import {
  StagehandClientActOptionsSchema,
  StagehandClientExtractOptionsSchema,
  StagehandClientInitParamsSchema,
  StagehandClientObserveOptionsSchema,
  type StagehandClientActOptions,
  type StagehandClientExtractOptions,
  type ResolvedStagehandClientLoggingConfig,
  type ResolvedStagehandClientInitParams,
  type StagehandClientInitParams,
  type StagehandClientObserveOptions,
} from "./clientSchemas.js";
import { CDPConnectionClosedError } from "./cdpClient.js";
import { STAGEHAND_EXTENSION_DIRECTORY_PATH } from "./extensionAssets.js";
import { STAGEHAND_SDK_CLIENT_INFO } from "./sdkIdentity.js";

type StagehandAdapters = {
  resolveBrowserSource?: (initParams: StagehandClientInitParams) => Promise<ResolvedBrowserSource>;
  connectRpcClient?: (options: RPCClientOptions) => Promise<RPCClient>;
};

const stagehandAdapters = new WeakMap<Stagehand, StagehandAdapters>();

type ProtocolExtractResult = import("../../protocol/types.js").ExtractResult;

export type ExtractResult<Schema extends z.ZodType> = Omit<ProtocolExtractResult, "data"> & {
  data: z.output<Schema>;
};

export class Stagehand {
  browserContext: BrowserContext | undefined;
  isInitialized = false;
  rpcClient: RPCClient | undefined;
  removeNotificationListener: (() => void) | undefined;
  removeClientLLMHandler: (() => void) | undefined;
  private resolvedBrowser: ResolvedBrowserSource | undefined;
  closePromise: Promise<void> | undefined;

  constructor(readonly initParams: StagehandClientInitParams) {}

  get context(): BrowserContext {
    if (!this.browserContext) {
      throw new Error("Stagehand is not initialized. Call stagehand.init() before using context.");
    }
    return this.browserContext;
  }

  get browser(): ResolvedBrowserSource {
    if (!this.resolvedBrowser) {
      throw new Error("Stagehand is not initialized. Call stagehand.init() before using browser.");
    }
    return this.resolvedBrowser;
  }

  get initialized(): boolean {
    return this.isInitialized;
  }

  async metrics(): Promise<StagehandMetrics> {
    return this.connectedRpcClient.send(StagehandMethods.stagehandMetrics, {});
  }

  async init(): Promise<void> {
    if (this.isInitialized) {
      return;
    }

    const clientInitParams = StagehandClientInitParamsSchema.parse(this.initParams);
    const adapters = stagehandAdapters.get(this) ?? {};
    const browser = await (adapters.resolveBrowserSource ?? resolveBrowserSource)(clientInitParams);
    this.resolvedBrowser = browser;

    try {
      const rpcClient = await (adapters.connectRpcClient ?? connectRPCClient)({
        cdpUrl: browser.cdpUrl,
        // TODO: Thread browser.cdpHeaders through CDP discovery and the WebSocket handshake.
        ...(browser.preloadedExtension
          ? { preloadedExtension: true as const }
          : { extensionDir: STAGEHAND_EXTENSION_DIRECTORY_PATH }),
        serviceWorkerUrlIncludes: "service-worker.js",
      });
      this.rpcClient = rpcClient;
      this.removeNotificationListener = rpcClient.onNotification((notification) =>
        handleStagehandNotification(notification, clientInitParams.logging),
      );
      if (clientInitParams.model && "generate" in clientInitParams.model) {
        this.removeClientLLMHandler = rpcClient.onRequest(
          StagehandMethods.llmGenerate,
          clientInitParams.model.generate,
        );
      }

      await rpcClient.send(
        StagehandMethods.stagehandInit,
        stagehandInitParamsForWorker(clientInitParams, browser, rpcClient),
      );
      this.browserContext = new BrowserContext(rpcClient);
    } catch (error) {
      this.removeClientLLMHandler?.();
      this.removeClientLLMHandler = undefined;
      this.removeNotificationListener?.();
      this.removeNotificationListener = undefined;
      this.rpcClient?.close();
      this.rpcClient = undefined;
      try {
        await this.closeBrowserSource();
      } catch (cleanupError) {
        throw new AggregateError(
          [error, cleanupError],
          "Stagehand initialization failed and browser cleanup also failed",
          { cause: error },
        );
      }
      throw error;
    }

    this.isInitialized = true;
    this.closePromise = undefined;
  }

  async act(instruction: string, options?: StagehandClientActOptions): Promise<ActResult>;
  async act(instruction: Action, options?: StagehandClientActOptions): Promise<ActResult>;
  async act(instruction: string | Action, options?: StagehandClientActOptions): Promise<ActResult> {
    const { page, ...protocolOptions } = StagehandClientActOptionsSchema.parse(options ?? {});
    const targetPage = page ?? (await this.context.activePage());
    if (!targetPage) throw new Error("Stagehand has no active page.");
    const response = await this.connectedRpcClient.send(StagehandMethods.stagehandAct, {
      pageId: targetPage.pageId,
      instruction,
      ...(options === undefined ? {} : { options: protocolOptions }),
    });

    return response;
  }

  async observe(
    instruction?: string,
    options?: StagehandClientObserveOptions,
  ): Promise<ObserveResult> {
    const { page, ...protocolOptions } = StagehandClientObserveOptionsSchema.parse(options ?? {});
    const targetPage = page ?? (await this.context.activePage());
    if (!targetPage) throw new Error("Stagehand has no active page.");
    const response = await this.connectedRpcClient.send(StagehandMethods.stagehandObserve, {
      pageId: targetPage.pageId,
      ...(instruction === undefined ? {} : { instruction }),
      ...(options === undefined ? {} : { options: protocolOptions }),
    });

    return response;
  }

  async extract<Schema extends z.ZodType>(
    instruction: string,
    schema: Schema,
    options?: StagehandClientExtractOptions,
  ): Promise<ExtractResult<Schema>> {
    const { page, ...protocolOptions } = StagehandClientExtractOptionsSchema.parse(options ?? {});
    const targetPage = page ?? (await this.context.activePage());
    if (!targetPage) throw new Error("Stagehand has no active page.");
    const jsonSchema = z.json().parse(z.toJSONSchema(schema));
    const response = await this.connectedRpcClient.send(StagehandMethods.stagehandExtract, {
      pageId: targetPage.pageId,
      instruction,
      schema: jsonSchema,
      ...(options === undefined ? {} : { options: protocolOptions }),
    });

    return {
      ...response,
      data: schema.parse(response.data),
    };
  }

  close(): Promise<void> {
    this.closePromise ??= (async () => {
      const context = this.browserContext;
      try {
        if (context) {
          try {
            await this.rpcClient?.send(StagehandMethods.stagehandClose, {});
          } catch (error) {
            if (!(error instanceof CDPConnectionClosedError)) throw error;
          }
        }
      } finally {
        this.removeClientLLMHandler?.();
        this.removeClientLLMHandler = undefined;
        this.removeNotificationListener?.();
        this.removeNotificationListener = undefined;
        this.rpcClient?.close();
        await this.closeBrowserSource();
        this.rpcClient = undefined;
        this.browserContext = undefined;
        this.isInitialized = false;
      }
    })();
    return this.closePromise;
  }

  private get connectedRpcClient(): RPCClient {
    if (!this.isInitialized || !this.rpcClient) {
      throw new Error("Stagehand is not initialized. Call stagehand.init() before using it.");
    }
    return this.rpcClient;
  }

  private async closeBrowserSource(): Promise<void> {
    const browser = this.resolvedBrowser;
    this.resolvedBrowser = undefined;
    if (!browser || browser.keepAlive) {
      return;
    }
    await browser.close?.();
  }
}

function stagehandInitParamsForWorker(
  initParams: ResolvedStagehandClientInitParams,
  resolvedBrowser: ResolvedBrowserSource,
  rpcClient: RPCClient,
) {
  const { browser, logging, model, ...protocolParams } = initParams;
  const protocolModel = model && "generate" in model ? { source: "client" as const } : model;

  if (browser.type === "browserbase" && !resolvedBrowser.browserbaseSessionId) {
    throw new Error("Resolved Browserbase source is missing its session ID");
  }
  if (!resolvedBrowser.residentBrowserConnection && !rpcClient.browserWebSocketDebuggerUrl) {
    throw new Error("The browser CDP WebSocket URL is unavailable");
  }

  return StagehandInitParamsSchema.parse({
    protocolVersion: STAGEHAND_PROTOCOL_VERSION,
    clientInfo: STAGEHAND_SDK_CLIENT_INFO,
    logLevel: logging.level,
    ...(resolvedBrowser.residentBrowserConnection
      ? {}
      : { browserCdpUrl: rpcClient.browserWebSocketDebuggerUrl }),
    ...protocolParams,
    ...(browser.type === "browserbase"
      ? {
          browser: {
            ...browser,
            sessionId: resolvedBrowser.browserbaseSessionId,
          },
        }
      : {}),
    ...(protocolModel === undefined ? {} : { model: protocolModel }),
  });
}

export function createStagehandWithClientForTest(client: RPCClient): Stagehand {
  return createStagehandWithDependenciesForTest(
    {
      browser: {
        type: "cdp",
        cdpUrl: "test://stagehand",
      },
    },
    {
      resolveBrowserSource: async () => ({
        cdpUrl: "test://stagehand",
        residentBrowserConnection: false,
        keepAlive: true,
      }),
      connectRpcClient: async () => client,
    },
  );
}

const LOG_LEVEL_PRIORITY = {
  debug: 10,
  info: 20,
  warn: 30,
  error: 40,
  off: Number.POSITIVE_INFINITY,
} as const;

function handleStagehandNotification(
  notification: StagehandRpcNotification,
  logging: ResolvedStagehandClientLoggingConfig,
): void {
  const log = notification.params;
  if (LOG_LEVEL_PRIORITY[log.level] < LOG_LEVEL_PRIORITY[logging.level]) return;

  process.stderr.write(renderStagehandLog(log, logging.format) + "\n");
  if (!logging.onLog) return;

  try {
    const result = logging.onLog(log);
    if (result instanceof Promise) {
      void result.catch(reportOnLogError);
    }
  } catch (error) {
    reportOnLogError(error);
  }
}

function renderStagehandLog(
  log: StagehandRpcNotification["params"],
  format: ResolvedStagehandClientLoggingConfig["format"],
): string {
  if (format === "json") return JSON.stringify(log);

  const data = Object.keys(log.data).length === 0 ? "" : ` ${JSON.stringify(log.data)}`;
  return `[stagehand] ${log.level.toUpperCase()} ${log.message}${data}`;
}

function reportOnLogError(error: unknown): void {
  const message = error instanceof Error ? error.message : String(error);
  process.stderr.write(`[stagehand] ERROR onLog callback failed: ${message}\n`);
}

export function createStagehandWithDependenciesForTest(
  initParams: StagehandClientInitParams,
  adapters: StagehandAdapters,
): Stagehand {
  const stagehand = new Stagehand(initParams);
  stagehandAdapters.set(stagehand, adapters);
  return stagehand;
}
